frame/support/src/traits.rs

```diff
@@ -1413,38 +1413,42 @@ impl<T> InitializeMembers<T> for () {
 	fn initialize_members(_: &[T]) {}
 }
 
-// A trait that is able to provide randomness.
-pub trait Randomness<Output> {
-	/// Get a "random" value
-	///
-	/// Being a deterministic blockchain, real randomness is difficult to come by. This gives you
-	/// something that approximates it. At best, this will be randomness which was
-	/// hard to predict a long time ago, but that has become easy to predict recently.
-	///
-	/// `subject` is a context identifier and allows you to get a
-	/// different result to other callers of this function; use it like
-	/// `random(&b"my context"[..])`.
-	fn random(subject: &[u8]) -> Output;
+/// A trait that is able to provide randomness.
+///
+/// Being a deterministic blockchain, real randomness is difficult to come by, different
+/// implementations of this trait will provide different security guarantees. At best,
+/// this will be randomness which was hard to predict a long time ago, but that has become
+/// easy to predict recently.
+pub trait Randomness<Output, BlockNumber> {
+	/// Get the most recently determined random seed, along with the time in the past
+	/// since when it was determinable by chain observers.
+	///
+	/// `subject` is a context identifier and allows you to get a different result to
+	/// other callers of this function; use it like `random(&b"my context"[..])`.
+	///
+	/// NOTE: The returned seed should only be used to distinguish commitments made before
+	/// the returned block number. If the block number is too early (i.e. commitments were
+	/// made afterwards), then ensure no further commitments may be made and repeatedly
+	/// call this on later blocks until the block number returned is later than the latest
+	/// commitment.
+	fn random(subject: &[u8]) -> (Output, BlockNumber);
 
 	/// Get the basic random seed.
 	///
-	/// In general you won't want to use this, but rather `Self::random` which allows you to give a
-	/// subject for the random result and whose value will be independently low-influence random
-	/// from any other such seeds.
-	fn random_seed() -> Output {
+	/// In general you won't want to use this, but rather `Self::random` which allows
+	/// you to give a subject for the random result and whose value will be
+	/// independently low-influence random from any other such seeds.
+	///
+	/// NOTE: The returned seed should only be used to distinguish commitments made before
+	/// the returned block number. If the block number is too early (i.e. commitments were
+	/// made afterwards), then ensure no further commitments may be made and repeatedly
+	/// call this on later blocks until the block number returned is later than the latest
+	/// commitment.
+	fn random_seed() -> (Output, BlockNumber) {
```

Comment:
isnt this pretty unique for a `Randomness` trait?

I feel the old `Randomness` trait was good, and this would be like `BlockInfluencedRandomness` or something....

Reply:
All randomness is relative like I explain in the text, so there is always some block number from before which all other computational inputs necessarily originate, or else the randomness is not fresh and hence non-random.  

We've three choices (1) relegate this fact purely to documentation like before, (2) support runtime asserts checking origin, or (3) somehow encode age using const generic. 

We should however discuss what this number means:  BABE VRF outputs would only be considered fresh with each block provided you trust the block producer somehow.  If you cannot trust block producers then BABE VRFs provide nothing better than `RandomnessFromTwoEpochsAgo`.  So which height do we put into the BAB VRF source?

I'd suggest we place the same height as `RandomnessFromTwoEpochsAgo` and simply explain that sometimes you can violate this limit.

Reply:
> Or do we expect all random number generators in the future to contain and be dependent on this kind of output?

We must support this API for now. In the future we might be able to have the RNG return a random number with the same block number (like with a randao) in which code using the Randomness won't need to delay after commitments are stopped. But for now (and a long while to come yet, probably) that's not the case so all code should support delay-after-commitments.

Reply:
I do think strong per block randomness looks way more realistic now thanks to https://github.com/kobigurk/aggregatable-dkg but..

If we ever produce strong per block randomness, then we'd use do so only as part of the BEEFY post-grandpa round.  This would tell us what happened due to the block, but only after the block was finalized, which likely suffices.

Reply:
> I feel the old Randomness trait was good, and this would be like BlockInfluencedRandomness or something....

Given the context ignoring it with a comment seems clearer to me: 

```
// We don't care about the freshness of the seed here:
let (seed, _) = T::Randomness::random();
```